src/qdisp/Executive.cc

```diff
@@ -108,25 +111,65 @@ Executive::Executive(ExecutiveConfig const& c, shared_ptr<MessageStore> const& m
     _secondsBetweenQMetaUpdates = chrono::seconds(_config.secondsBetweenChunkUpdates);
     _setup();
     _setupLimit();
+    qdisp::CzarStats::get()->addQuery();
 }
 
 Executive::~Executive() {
+    qdisp::CzarStats::get()->deleteQuery();
+    qdisp::CzarStats::get()->deleteJobs(_incompleteJobs.size());
     // Real XrdSsiService objects are unowned, but mocks are allocated in _setup.
     delete dynamic_cast<XrdSsiServiceMock*>(_xrdSsiService);
+    if (_asyncTimer != nullptr) {
+        _asyncTimer->cancel();
+        qdisp::CzarStats::get()->untrackQueryProgress(_id);
+    }
 }
 
 Executive::Ptr Executive::create(ExecutiveConfig const& c, shared_ptr<MessageStore> const& ms,
                                  SharedResources::Ptr const& sharedResources,
                                  shared_ptr<qmeta::QStatus> const& qMeta,
-                                 shared_ptr<qproc::QuerySession> const& querySession) {
-    Executive::Ptr exec(new Executive(c, ms, sharedResources, qMeta,
-                                      querySession));  // make_shared dislikes private constructor.
-    return exec;
+                                 shared_ptr<qproc::QuerySession> const& querySession,
+                                 boost::asio::io_service& asioIoService) {
+    LOGS(_log, LOG_LVL_DEBUG, "Executive::" << __func__);
+    Executive::Ptr ptr(new Executive(c, ms, sharedResources, qMeta, querySession));
+
+    // Start the query progress monitoring timer (if enabled). The query status
+    // will be sampled on each expiration event of the timer. Note that the timer
+    // gets restarted automatically for as long as the context (the current
+    // Executive object) still exists.
+    //
+    // IMPORTANT: The weak pointer dependency (unlike the regular shared pointer)
+    // is required here to allow destroying the Executive object without explicitly
+    // stopping the timer.
+    auto const czarStatsUpdateIvalSec = cconfig::CzarConfig::instance()->czarStatsUpdateIvalSec();
+    if (czarStatsUpdateIvalSec > 0) {
+        ptr->_asyncTimer = util::AsyncTimer::create(
+                asioIoService, std::chrono::milliseconds(czarStatsUpdateIvalSec * 1000),
+                [self = std::weak_ptr<Executive>(ptr)](auto expirationIvalMs) -> bool {
+                    auto ptr = self.lock();
+                    LOGS(_log, LOG_LVL_DEBUG,
+                         "Executive::" << __func__ << " expirationIvalMs: " << expirationIvalMs.count()
+                                       << " ms");
+                    if (ptr != nullptr) {
+                        ptr->_updateStats();
+                        return true;
+                    }
+                    return false;
+                });
+        ptr->_asyncTimer->start();
+    }
+    return ptr;
+}
```

Comment:
Yes, I see value in your proposal - the query states _may_ be captured in sync (one timestamp for all). Though, it may not matter much from the monitoring point of view. The time series plotting tool that I'm using works well with async series. I might reconsider this if I were to switch to the sync regime.

In the current design, a chain of timer expiration events starts at the construction time of each Executive. The chain finishes when the Executive gets distracted. This mechanism acts as the light-weight observer of the evolving state of the Executive. The callbacks are executed within the dedicated thread pool associated with the ASIO's `io_service` object. The design decouples thread management (for timers) from the timer management (within the Executive). It also decouples instances of Executives. Each one has its own independent "observer" (the timer).